query.go

```diff
@@ -3,317 +3,362 @@ package dht
+		closer := q.localPeers.Add(np.ID)
+		foundCloserPeer = foundCloserPeer || closer
+	}
+
+	if q.stopFn(q.localPeers) {
+		q.cancel()
+	}
+	return &queryResult{
+		success:         true,
+		foundCloserPeer: foundCloserPeer,
 	}
 }
 
-func (r *dhtQueryRunner) dialPeer(ctx context.Context, p peer.ID) error {
+func (dht *IpfsDHT) dialPeer(ctx context.Context, p peer.ID) error {
 	// short-circuit if we're already connected.
-	if r.query.dht.host.Network().Connectedness(p) == network.Connected {
+	if dht.host.Network().Connectedness(p) == network.Connected {
 		return nil
 	}
 
 	logger.Debug("not connected. dialing.")
-	notif.PublishQueryEvent(r.runCtx, &notif.QueryEvent{
-		Type: notif.DialingPeer,
+	routing.PublishQueryEvent(ctx, &routing.QueryEvent{
+		Type: routing.DialingPeer,
 		ID:   p,
 	})
 
 	pi := peer.AddrInfo{ID: p}
-	if err := r.query.dht.host.Connect(ctx, pi); err != nil {
+	if err := dht.host.Connect(ctx, pi); err != nil {
 		logger.Debugf("error connecting: %s", err)
-		notif.PublishQueryEvent(r.runCtx, &notif.QueryEvent{
-			Type:  notif.QueryError,
+		routing.PublishQueryEvent(ctx, &routing.QueryEvent{
+			Type:  routing.QueryError,
 			Extra: err.Error(),
 			ID:    p,
 		})
 
-		// This peer is dropping out of the race.
-		r.peersRemaining.Decrement(1)
 		return err
 	}
 	logger.Debugf("connected. dial success.")
 	return nil
 }
 
-func (r *dhtQueryRunner) queryPeer(proc process.Process, p peer.ID) {
-	// ok let's do this!
-
-	// create a context from our proc.
-	ctx := ctxproc.OnClosingContext(proc)
-
-	// make sure we do this when we exit
-	defer func() {
-		// signal we're done processing peer p
-		r.peersRemaining.Decrement(1)
-		r.rateLimit <- struct{}{}
-	}()
-
-	// finally, run the query against this peer
-	res, err := r.query.qfunc(ctx, p)
-
-	r.peersQueried.Add(p)
-
-	if err != nil {
-		logger.Debugf("ERROR worker for: %v %v", p, err)
-	} else if res.success {
-		logger.Debugf("SUCCESS worker for: %v %s", p, res)
-		r.Lock()
-		r.result = res
-		r.Unlock()
-		if res.peer != nil {
-			r.query.dht.peerstore.AddAddrs(res.peer.ID, res.peer.Addrs, pstore.TempAddrTTL)
-		}
-		go r.proc.Close() // signal to everyone that we're done.
-		// must be async, as we're one of the children, and Close blocks.
-
-	} else if len(res.closerPeers) > 0 {
-		logger.Debugf("PEERS CLOSER -- worker for: %v (%d closer peers)", p, len(res.closerPeers))
-		for _, next := range res.closerPeers {
-			if next.ID == r.query.dht.self { // don't add self.
-				logger.Debugf("PEERS CLOSER -- worker for: %v found self", p)
-				continue
-			}
-
-			// add their addresses to the dialer's peerstore
-			r.query.dht.peerstore.AddAddrs(next.ID, next.Addrs, pstore.TempAddrTTL)
-			r.addPeerToQuery(next.ID)
-			logger.Debugf("PEERS CLOSER -- worker for: %v added %v (%v)", p, next.ID, next.Addrs)
+// Equal tells whether a and b contain the same elements.
+// A nil argument is equivalent to an empty slice.
+func peerSlicesEqual(a, b []peer.ID) bool {
```

Comment:
not getting used.

Reply:
it's being used by the unused query functions you pointed out above.

Reply:
function and relevant query functions removed